

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Application of PNet to TCGA BRCA &mdash; PNetTorch</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
      <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
      <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css?v=be8a1c11" />
      <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
      <link rel="stylesheet" type="text/css" href="../_static/design-style.1e8bd061cd6da7fc9cf755528e8ffc24.min.css?v=0a3b3ea7" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
      <script src="../_static/doctools.js?v=9a2dae69"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
      <script src="../_static/copybutton.js?v=f281be69"></script>
      <script src="../_static/scripts/sphinx-book-theme.js"></script>
      <script>let toggleHintShow = 'Click to show';</script>
      <script>let toggleHintHide = 'Click to hide';</script>
      <script>let toggleOpenOnPrint = 'true';</script>
      <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
      <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
      <script src="../_static/design-tabs.js?v=36754332"></script>
      <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
      <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
      <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
      <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="PNet Module" href="MAIN/Pnet.html" />
    <link rel="prev" title="A replication of the results on the Prostate Dataset from PNet" href="PNet_Prostate.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../intro.html" class="icon icon-home">
            Python
              <img src="../_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="PNet_Prostate.html">A replication of the results on the Prostate Dataset from PNet </a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Application of PNet to TCGA BRCA</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">PNet Modules</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="MAIN/Pnet.html">PNet Module</a></li>
<li class="toctree-l1"><a class="reference internal" href="MAIN/reactome.html">Reactome Pathway Network Module</a></li>
<li class="toctree-l1"><a class="reference internal" href="MAIN/train.html">Training Module</a></li>
<li class="toctree-l1"><a class="reference internal" href="MAIN/utils.html">Utility Functions</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../intro.html">Python</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../intro.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Application of PNet to TCGA BRCA</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/Barry8197/PNetBook/blob/master/_build/html/_sources/PNetTorch/PNet_BRCA.ipynb" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section class="tex2jax_ignore mathjax_ignore" id="application-of-pnet-to-tcga-brca">
<h1>Application of PNet to TCGA BRCA<a class="headerlink" href="#application-of-pnet-to-tcga-brca" title="Link to this heading"></a></h1>
<p>The <a class="reference external" href="https://portal.gdc.cancer.gov/projects/TCGA-BRCA">TCGA-BRCA</a> project comes from <a class="reference external" href="https://portal.gdc.cancer.gov/">The Cancer Genome Atlas</a>. This dataset consists of clinical, genomic and more data types. In this tutorial we explore the use of two data types :</p>
<ul class="simple">
<li><p>mRNA</p></li>
<li><p>Reverse Phase Protein Array (RPPA)</p></li>
</ul>
<p>BRCA breast cancer refers to breast cancer that is linked with mutations in either the BRCA1 or BRCA2 genes. These genes—BRCA1 (Breast Cancer gene one) and BRCA2 (Breast Cancer gene two)—are important in the body’s mechanism for repairing DNA damage. Normally, they help ensure the stability of the cell’s genetic material and prevent uncontrolled cell growth. Mutation in these genes can lead to damaged DNA not being properly repaired, leading to further genetic alterations that can cause cancer.</p>
<section id="key-points-about-brca-related-breast-cancer">
<h2>Key Points about BRCA-Related Breast Cancer:<a class="headerlink" href="#key-points-about-brca-related-breast-cancer" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p><strong>Hereditary Cancer</strong>: <br>
Mutations in these genes are hereditary (inherited from a parent), making BRCA1 or BRCA2 mutations a strong risk factor for developing breast cancer, as well as other types of cancer, such as ovarian cancer.</p></li>
<li><p><strong>Increased Risk</strong>: <br>
Individuals with BRCA1 or BRCA2 gene mutations have a higher risk of developing breast cancer at a younger age than those without such mutations. The lifetime risk of breast cancer in women with a BRCA1 mutation is estimated to be between 55% and 65%, and for BRCA2 mutations, it is around 45%.</p></li>
</ul>
<p>In this tutorial we will aim to predict 5 different subtypes of BRCA breast cancer outlined by the TCGA. These subtypes are :</p>
<ul class="simple">
<li><p>Luminal A</p></li>
<li><p>Luminal B</p></li>
<li><p>HER2</p></li>
<li><p>Basal</p></li>
<li><p>Normal</p></li>
</ul>
<p>These subtypes are defined by an array of 50 gene expression signatures making this problem a suitable task for validating a genomic algorithm like PNet</p>
</section>
<section id="import-packages-and-pnettorch-modules">
<h2>Import Packages and PNetTorch Modules<a class="headerlink" href="#import-packages-and-pnettorch-modules" title="Link to this heading"></a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">sys</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span> <span class="p">,</span> <span class="s1">&#39;./MAIN/&#39;</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">reactome</span> <span class="kn">import</span> <span class="n">ReactomeNetwork</span>
<span class="kn">from</span> <span class="nn">Pnet</span> <span class="kn">import</span> <span class="n">MaskedLinear</span> <span class="p">,</span> <span class="n">PNET</span>
<span class="kn">from</span> <span class="nn">train</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">utils</span> <span class="kn">import</span> <span class="n">numpy_array_to_one_hot</span><span class="p">,</span> <span class="n">get_gpu_memory</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">gc</span>
<span class="kn">import</span> <span class="nn">pickle</span>

<span class="o">%</span><span class="k">load_ext</span> autoreload
<span class="o">%</span><span class="k">autoreload</span> 2
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;frozen importlib._bootstrap&gt;:228: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject
</pre></div>
</div>
</div>
</div>
</section>
<section id="import-data">
<h2>Import Data<a class="headerlink" href="#import-data" title="Link to this heading"></a></h2>
<p>Data can be downloaded from <a class="reference external" href="https://portal.gdc.cancer.gov/projects/TCGA-BRCA">https://portal.gdc.cancer.gov/projects/TCGA-BRCA</a></p>
<p>mRNA data was previously processed using DESeq packages for gene expression normalisation and gene expression filtering.</p>
<p>RPPA data was processed by normalisation</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">datMeta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([])</span>
<span class="n">datExpr</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([])</span>

<span class="k">for</span> <span class="n">mod</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;mRNA&#39;</span> <span class="p">,</span> <span class="s1">&#39;RPPA&#39;</span>  <span class="p">]</span> <span class="p">:</span> 
    
    <span class="c1"># Reading from a pickle file</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;./data/</span><span class="si">{</span><span class="n">mod</span><span class="si">}</span><span class="s1">_processed.pkl&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>  
        <span class="n">data</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;datExpr&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;datMeta&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="p">:</span> 
        <span class="n">data</span><span class="p">[</span><span class="s1">&#39;datExpr&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;datExpr&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">T</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">datMeta</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="p">:</span> 
        <span class="n">datExpr</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;datExpr&#39;</span><span class="p">]</span>
        <span class="n">datMeta</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;datMeta&#39;</span><span class="p">]</span>
    <span class="k">else</span> <span class="p">:</span> 
        <span class="n">datExpr</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">datExpr</span> <span class="p">,</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;datExpr&#39;</span><span class="p">]</span> <span class="p">,</span> <span class="n">right_index</span> <span class="o">=</span> <span class="kc">True</span> <span class="p">,</span> <span class="n">left_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">datMeta</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">datMeta</span> <span class="p">,</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;datMeta&#39;</span><span class="p">]</span>  <span class="p">,</span> <span class="n">left_index</span><span class="o">=</span><span class="kc">True</span> <span class="p">,</span> <span class="n">right_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        
<span class="n">datMeta</span> <span class="o">=</span> <span class="n">datMeta</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;paper_BRCA_Subtype_PAM50_y&#39;</span> <span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">datMeta</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;paper_BRCA_Subtype_PAM50&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">datExpr</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(840, 30459)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">datMeta</span> <span class="o">=</span> <span class="n">datMeta</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">datExpr</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="run-pnettorch-on-data">
<h2>Run PNetTorch on Data<a class="headerlink" href="#run-pnettorch-on-data" title="Link to this heading"></a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">torch</span>
<span class="kn">import</span> <span class="nn">torch.nn</span> <span class="k">as</span> <span class="nn">nn</span>
<span class="kn">import</span> <span class="nn">torch.nn.functional</span> <span class="k">as</span> <span class="nn">F</span>

<span class="n">x</span> <span class="o">=</span> <span class="n">datExpr</span><span class="o">.</span><span class="n">values</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">datMeta</span><span class="p">[</span><span class="s1">&#39;paper_BRCA_Subtype_PAM50&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">codes</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span>

<span class="c1"># List of cancer genes was taken from the PNet paper dataset</span>
<span class="n">genes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;./paper_data/genes/cancer_genes.txt&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># Highly recommend running on GPU or in HPC environment</span>
<span class="n">device</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">device</span><span class="p">(</span><span class="s1">&#39;cuda&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">is_available</span><span class="p">()</span> <span class="k">else</span> <span class="n">torch</span><span class="o">.</span><span class="n">device</span><span class="p">(</span><span class="s1">&#39;cpu&#39;</span><span class="p">)</span>
<span class="c1">#device = torch.device(&#39;mps&#39;)</span>

<span class="c1"># Build network to obtain gene and pathway relationships</span>
<span class="n">net</span> <span class="o">=</span> <span class="n">ReactomeNetwork</span><span class="p">(</span><span class="n">genes_of_interest</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">genes</span><span class="p">[</span><span class="s1">&#39;genes&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">))</span> <span class="p">,</span> <span class="n">n_levels</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>

<span class="c1"># Build PNet model</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">PNET</span><span class="p">(</span><span class="n">reactome_network</span><span class="o">=</span><span class="n">net</span><span class="p">,</span> <span class="n">input_dim</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="p">,</span> <span class="n">output_dim</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">y</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> 
                  <span class="n">activation</span> <span class="o">=</span> <span class="n">nn</span><span class="o">.</span><span class="n">ReLU</span> <span class="p">,</span> <span class="n">dropout</span><span class="o">=</span><span class="mf">0.1</span> <span class="p">,</span> <span class="n">filter_pathways</span><span class="o">=</span><span class="kc">False</span> <span class="p">,</span> <span class="n">input_layer_mask</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">model</span><span class="p">)</span>

<span class="c1"># Convert classifications to One Hot Encodings</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">numpy_array_to_one_hot</span><span class="p">(</span><span class="n">y</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>

<span class="c1"># Train Model</span>
<span class="n">model</span> <span class="p">,</span> <span class="n">train_loader</span> <span class="p">,</span> <span class="n">test_loader</span> <span class="o">=</span> <span class="n">train</span><span class="p">(</span><span class="n">model</span> <span class="p">,</span> <span class="n">x</span> <span class="p">,</span> <span class="n">y</span> <span class="p">,</span> <span class="n">device</span> <span class="p">,</span> <span class="n">batch_size</span> <span class="o">=</span> <span class="mi">1024</span> <span class="p">,</span> <span class="n">lr_decay</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
                                                 <span class="n">num_epochs</span><span class="o">=</span><span class="mi">200</span> <span class="p">,</span> <span class="n">learning_rate</span><span class="o">=</span><span class="mf">0.001</span> <span class="p">,</span> <span class="n">sparse</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="c1"># Evaluate</span>
<span class="n">metrics</span> <span class="o">=</span> <span class="n">evaluate</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">test_loader</span><span class="p">,</span> <span class="n">device</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">metrics</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>PNET(
  (dropout): Dropout(p=0.1, inplace=False)
  (layers): ModuleList(
    (0): Linear(in_features=30459, out_features=723, bias=True)
    (1): MaskedLinear(in_features=723, out_features=1397, bias=True)
    (2): MaskedLinear(in_features=1397, out_features=1066, bias=True)
    (3): MaskedLinear(in_features=1066, out_features=447, bias=True)
    (4): MaskedLinear(in_features=447, out_features=147, bias=True)
    (5): MaskedLinear(in_features=147, out_features=26, bias=True)
  )
  (skip): ModuleList(
    (0): Linear(in_features=723, out_features=5, bias=True)
    (1): Linear(in_features=1397, out_features=5, bias=True)
    (2): Linear(in_features=1066, out_features=5, bias=True)
    (3): Linear(in_features=447, out_features=5, bias=True)
    (4): Linear(in_features=147, out_features=5, bias=True)
    (5): Linear(in_features=26, out_features=5, bias=True)
  )
  (act_layers): ModuleList(
    (0-5): 6 x ReLU()
  )
)
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 200: 100%|██████████| 200/200 [00:18&lt;00:00, 10.76epoch/s, loss=1.91]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Training complete.
              precision    recall  f1-score   support

           0     0.9231    0.8571    0.8889        14
           1     1.0000    0.6250    0.7692         8
           2     0.7000    1.0000    0.8235        42
           3     1.0000    0.3529    0.5217        17
           4     0.0000    0.0000    0.0000         3

    accuracy                         0.7738        84
   macro avg     0.7246    0.5670    0.6007        84
weighted avg     0.8015    0.7738    0.7388        84
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>C:\Users\Barry\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1497: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f&quot;{metric.capitalize()} is&quot;, len(result))
C:\Users\Barry\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1497: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f&quot;{metric.capitalize()} is&quot;, len(result))
C:\Users\Barry\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1497: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f&quot;{metric.capitalize()} is&quot;, len(result))
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">actuals</span><span class="p">,</span> <span class="n">predictions</span> <span class="o">=</span> <span class="n">get_predictions</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="s1">&#39;cpu&#39;</span><span class="p">),</span> <span class="n">test_loader</span><span class="p">)</span>

<span class="n">cm</span> <span class="o">=</span> <span class="n">plot_confusion_matrix</span><span class="p">(</span><span class="n">actuals</span> <span class="p">,</span> <span class="n">predictions</span><span class="p">)</span>

<span class="n">actuals</span><span class="p">,</span> <span class="n">probs</span> <span class="o">=</span> <span class="n">get_probabilities</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="s1">&#39;cpu&#39;</span><span class="p">),</span> <span class="n">test_loader</span><span class="p">)</span>
<span class="c1"># Assuming `probs` has been calculated as the probability outcomes for class 1, if not implement it</span>
<span class="n">auc</span> <span class="o">=</span> <span class="n">roc_auc_score</span><span class="p">(</span><span class="n">actuals</span><span class="p">,</span> <span class="n">probs</span> <span class="p">,</span> <span class="n">multi_class</span><span class="o">=</span><span class="s1">&#39;ovr&#39;</span><span class="p">)</span>

<span class="n">roc</span> <span class="o">=</span> <span class="n">plot_roc_curve</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">actuals</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">probs</span><span class="p">)</span> <span class="p">,</span> <span class="n">n_classes</span> <span class="o">=</span> <span class="mi">5</span> <span class="p">,</span> <span class="n">target_names</span> <span class="o">=</span> <span class="n">datMeta</span><span class="p">[</span><span class="s1">&#39;paper_BRCA_Subtype_PAM50&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/b83663b89f9b4224f0d883524eaa50d8fe9623499c72cfc4538eb40c25434f4a.png" src="../_images/b83663b89f9b4224f0d883524eaa50d8fe9623499c72cfc4538eb40c25434f4a.png" />
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Micro-averaged One-vs-Rest ROC AUC score:
0.91
</pre></div>
</div>
<img alt="../_images/ac9b221c40e0635dde73ef52ffe0e6800162041bc43a2855ff1e12cc71f11ecd.png" src="../_images/ac9b221c40e0635dde73ef52ffe0e6800162041bc43a2855ff1e12cc71f11ecd.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">non_zero_sum</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">for</span> <span class="n">param</span> <span class="ow">in</span> <span class="n">model</span><span class="o">.</span><span class="n">parameters</span><span class="p">():</span>
    <span class="n">non_zero_elements</span> <span class="o">=</span> <span class="p">(</span><span class="n">param</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span> <span class="c1"># filter non-zero elements</span>
    <span class="n">non_zero_sum</span> <span class="o">+=</span> <span class="n">non_zero_elements</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>  <span class="c1"># sum them up</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sum of non-zero parameters:&quot;</span><span class="p">,</span> <span class="n">non_zero_sum</span><span class="o">.</span><span class="n">item</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Sum of non-zero parameters: 22056521
</pre></div>
</div>
</div>
</div>
<section id="functions-for-checking-and-clearing-gpu-memory-if-required">
<h3>Functions for checking and clearing GPU memory if required<a class="headerlink" href="#functions-for-checking-and-clearing-gpu-memory-if-required" title="Link to this heading"></a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">get_gpu_memory</span><span class="p">()</span>
<span class="c1">#del model , train_loader , test_loader</span>
<span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">reset_accumulated_memory_stats</span><span class="p">()</span>
<span class="n">gc</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span>
<span class="n">torch</span><span class="o">.</span><span class="n">cuda</span><span class="o">.</span><span class="n">empty_cache</span><span class="p">()</span>
<span class="n">get_gpu_memory</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Total = 6.4Gb 	 Reserved = 0.7Gb 	 Allocated = 0.0Gb
Total = 6.4Gb 	 Reserved = 0.0Gb 	 Allocated = 0.0Gb
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="interpretability">
<h2>Interpretability<a class="headerlink" href="#interpretability" title="Link to this heading"></a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">interpret</span> <span class="kn">import</span> <span class="n">interpret</span> <span class="p">,</span> <span class="n">evaluate_interpret_save</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Extract all data and targets using list comprehensions</span>
<span class="n">data_batches</span><span class="p">,</span> <span class="n">target_batches</span> <span class="o">=</span> <span class="nb">zip</span><span class="p">(</span><span class="o">*</span><span class="p">[(</span><span class="n">data</span><span class="p">,</span> <span class="n">target</span><span class="p">)</span> <span class="k">for</span> <span class="n">data</span><span class="p">,</span> <span class="n">target</span> <span class="ow">in</span> <span class="n">test_loader</span><span class="p">])</span>

<span class="c1"># Convert lists of batches into single tensors</span>
<span class="n">test_data</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">cat</span><span class="p">(</span><span class="n">data_batches</span><span class="p">,</span> <span class="n">dim</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">test_targets</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">cat</span><span class="p">(</span><span class="n">target_batches</span> <span class="p">,</span> <span class="n">dim</span> <span class="o">=</span><span class="mi">0</span><span class="p">)</span>

<span class="c1"># Assign the model features to the columns</span>
<span class="n">model</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="n">datExpr</span><span class="o">.</span><span class="n">columns</span>

<span class="n">model_importances</span> <span class="o">=</span> <span class="n">interpret</span><span class="p">(</span><span class="n">model</span> <span class="p">,</span> <span class="n">test_data</span> <span class="p">,</span> <span class="n">savedir</span><span class="o">=</span><span class="s1">&#39;None&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Save Path Not Found - Plots will not be saved
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>C:\Users\Barry\anaconda3\lib\site-packages\captum\_utils\gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
C:\Users\Barry\anaconda3\lib\site-packages\captum\attr\_core\deep_lift.py:304: UserWarning: Setting forward, backward hooks and attributes on non-linear
               activations. The hooks and attributes will be removed
            after the attribution is finished
  warnings.warn(
</pre></div>
</div>
<img alt="../_images/a78c14b881fb0d0ec4e0b670d04eba692090c3ab24c45ed8233b18bd233e9303.png" src="../_images/a78c14b881fb0d0ec4e0b670d04eba692090c3ab24c45ed8233b18bd233e9303.png" />
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Linear(in_features=30459, out_features=723, bias=True)
MaskedLinear(in_features=723, out_features=1397, bias=True)
MaskedLinear(in_features=1397, out_features=1066, bias=True)
MaskedLinear(in_features=1066, out_features=447, bias=True)
MaskedLinear(in_features=447, out_features=147, bias=True)
MaskedLinear(in_features=147, out_features=26, bias=True)
</pre></div>
</div>
<img alt="../_images/e6dcfb601056db2a473d44e63116bb2efff18116196fc3459f34926af4465b8b.png" src="../_images/e6dcfb601056db2a473d44e63116bb2efff18116196fc3459f34926af4465b8b.png" />
<img alt="../_images/fd6a0430b0c8709c9894739effd2e1f1858e1bcaf36255fb03d7482ede21b114.png" src="../_images/fd6a0430b0c8709c9894739effd2e1f1858e1bcaf36255fb03d7482ede21b114.png" />
<img alt="../_images/54a4b17d63eb19c85ee8902f55f647c56eb548356047b3f9644958dd7c74fcf3.png" src="../_images/54a4b17d63eb19c85ee8902f55f647c56eb548356047b3f9644958dd7c74fcf3.png" />
<img alt="../_images/3c6b8c387cebbb0eb15e0a9cc44637e7a7357cf4f4b0494ccb45b4501cb19072.png" src="../_images/3c6b8c387cebbb0eb15e0a9cc44637e7a7357cf4f4b0494ccb45b4501cb19072.png" />
<img alt="../_images/0b9bc24011a46156406b27211decc1aace59a74966d37140d6533a45aea9e874.png" src="../_images/0b9bc24011a46156406b27211decc1aace59a74966d37140d6533a45aea9e874.png" />
<img alt="../_images/5be3a776b4daeb17bbf2dfad2c27abfdb99c34aecfc04c32b2f1d55e10d7cfb0.png" src="../_images/5be3a776b4daeb17bbf2dfad2c27abfdb99c34aecfc04c32b2f1d55e10d7cfb0.png" />
</div>
</div>
</section>
<section id="generate-and-save-metrics-importance">
<h2>Generate and Save Metrics + Importance<a class="headerlink" href="#generate-and-save-metrics-importance" title="Link to this heading"></a></h2>
<p>Single function to generate and save all model metrics and importances</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">evaluate_interpret_save</span><span class="p">(</span><span class="n">model</span> <span class="p">,</span> <span class="n">test_loader</span> <span class="p">,</span> <span class="s1">&#39;./Output/Prostate&#39;</span> <span class="p">,</span> <span class="n">n_classes</span> <span class="o">=</span> <span class="mi">5</span> <span class="p">,</span> <span class="n">target_names</span> <span class="o">=</span> <span class="n">datMeta</span><span class="p">[</span><span class="s1">&#39;paper_BRCA_Subtype_PAM50&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/b83663b89f9b4224f0d883524eaa50d8fe9623499c72cfc4538eb40c25434f4a.png" src="../_images/b83663b89f9b4224f0d883524eaa50d8fe9623499c72cfc4538eb40c25434f4a.png" />
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AUC Score: 0.7736424106412905
Micro-averaged One-vs-Rest ROC AUC score:
0.91
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>C:\Users\Barry\anaconda3\lib\site-packages\captum\_utils\gradient.py:57: UserWarning: Input Tensor 0 did not already require gradients, required_grads has been set automatically.
  warnings.warn(
C:\Users\Barry\anaconda3\lib\site-packages\captum\attr\_core\deep_lift.py:304: UserWarning: Setting forward, backward hooks and attributes on non-linear
               activations. The hooks and attributes will be removed
            after the attribution is finished
  warnings.warn(
</pre></div>
</div>
<img alt="../_images/ac9b221c40e0635dde73ef52ffe0e6800162041bc43a2855ff1e12cc71f11ecd.png" src="../_images/ac9b221c40e0635dde73ef52ffe0e6800162041bc43a2855ff1e12cc71f11ecd.png" />
<img alt="../_images/a78c14b881fb0d0ec4e0b670d04eba692090c3ab24c45ed8233b18bd233e9303.png" src="../_images/a78c14b881fb0d0ec4e0b670d04eba692090c3ab24c45ed8233b18bd233e9303.png" />
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Linear(in_features=30459, out_features=723, bias=True)
MaskedLinear(in_features=723, out_features=1397, bias=True)
MaskedLinear(in_features=1397, out_features=1066, bias=True)
MaskedLinear(in_features=1066, out_features=447, bias=True)
MaskedLinear(in_features=447, out_features=147, bias=True)
MaskedLinear(in_features=147, out_features=26, bias=True)
</pre></div>
</div>
<img alt="../_images/e6dcfb601056db2a473d44e63116bb2efff18116196fc3459f34926af4465b8b.png" src="../_images/e6dcfb601056db2a473d44e63116bb2efff18116196fc3459f34926af4465b8b.png" />
<img alt="../_images/fd6a0430b0c8709c9894739effd2e1f1858e1bcaf36255fb03d7482ede21b114.png" src="../_images/fd6a0430b0c8709c9894739effd2e1f1858e1bcaf36255fb03d7482ede21b114.png" />
<img alt="../_images/54a4b17d63eb19c85ee8902f55f647c56eb548356047b3f9644958dd7c74fcf3.png" src="../_images/54a4b17d63eb19c85ee8902f55f647c56eb548356047b3f9644958dd7c74fcf3.png" />
<img alt="../_images/3c6b8c387cebbb0eb15e0a9cc44637e7a7357cf4f4b0494ccb45b4501cb19072.png" src="../_images/3c6b8c387cebbb0eb15e0a9cc44637e7a7357cf4f4b0494ccb45b4501cb19072.png" />
<img alt="../_images/0b9bc24011a46156406b27211decc1aace59a74966d37140d6533a45aea9e874.png" src="../_images/0b9bc24011a46156406b27211decc1aace59a74966d37140d6533a45aea9e874.png" />
<img alt="../_images/5be3a776b4daeb17bbf2dfad2c27abfdb99c34aecfc04c32b2f1d55e10d7cfb0.png" src="../_images/5be3a776b4daeb17bbf2dfad2c27abfdb99c34aecfc04c32b2f1d55e10d7cfb0.png" />
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./PNetTorch"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="PNet_Prostate.html" class="btn btn-neutral float-left" title="A replication of the results on the Prostate Dataset from PNet" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="MAIN/Pnet.html" class="btn btn-neutral float-right" title="PNet Module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>